src/harvester/geo_network_q.rs

```diff
@@ -0,0 +1,193 @@
+use anyhow::Result;
+use cap_std::fs::Dir;
+use futures_util::stream::{iter, StreamExt};
+use quick_xml::de::from_slice;
+use reqwest::Client;
+use serde::Deserialize;
+
+use crate::{
+    dataset::Dataset,
+    harvester::{with_retry, write_dataset, Source},
+};
+
+pub async fn harvest(dir: &Dir, client: &Client, source: &Source) -> Result<(usize, usize, usize)> {
+    let entries = source.batch_size;
+
+    let (count, results, errors) = fetch_datasets(dir, client, source, true, 1, entries).await?;
+    tracing::info!("Harvesting {} datasets", count);
+
+    let requests = (count + entries - 1) / entries;
+    let from = (1..requests).map(|request| 1 + request * entries);
+    let to = from.clone().map(|from| from + entries - 1);
+
+    iter(from.zip(to))
+        .map(|(from, to)| fetch_datasets(dir, client, source, false, from, to))
+        .buffer_unordered(source.concurrency)
+        .fold(
+            (results, errors),
+            |(mut results, mut errors), res| async move {
+                match res {
+                    Ok((_count, results1, errors1)) => {
+                        results += results1;
+                        errors += errors1;
+                    }
+                    Err(err) => {
+                        tracing::error!("{:#}", err);
+
+                        errors += 1;
+                    }
+                }
+
+                (results, errors)
+            },
+        )
+        .await;
+
+    Ok((count, results, errors))
+}
+
+#[tracing::instrument(skip(dir, client, source))]
+async fn fetch_datasets(
+    dir: &Dir,
+    client: &Client,
+    source: &Source,
+    summary: bool,
+    from: usize,
+    to: usize,
+) -> Result<(usize, usize, usize)> {
+    tracing::debug!("Fetching datasets from {} to {}", from, to);
+
+    let response = with_retry(|| async {
+        let body = client
+            .get(source.url.clone())
+            .query(&[
+                ("fast", "false"),
+                ("buildSummary", &summary.to_string()),
+                ("from", &from.to_string()),
+                ("to", &to.to_string()),
+            ])
+            .send()
+            .await?
+            .error_for_status()?
+            .bytes()
+            .await?;
+
+        let response: SearchResults = from_slice(&body)?;
+
+        Ok(response)
+    })
+    .await?;
+
+    let count = if let Some(summary) = response.summary {
+        summary.count.parse()?
+    } else {
+        0
+    };
+
+    let results = response.metadata.len();
+    let mut errors = 0;
+
+    for entry in response.metadata {
+        if let Err(err) = translate_dataset(dir, source, entry).await {
+            tracing::error!("{:#}", err);
```

Comment:
I would not have expected translate_dataset to contain the write_dataset function. Maybe we find a more transparent naming?

Reply:
translate_write_dataset?

Reply:
I agree that the naming is unfortunate and will try to think of a better variant.